[8.0][FIX][base_export_manager] Fix "Expected singleton" bug.

yajo opened the pull request:
If you had a field that got translated in more than 1 addon, you'd possibly get
to this error:

```
  File "/opt/odoo/0079_ahk_openerp/oca/base_export_manager/models/ir_exports_line.py", line 105, in _compute_label
    field.name)),
  File "/opt/odoo/common/openerp/v8/openerp/fields.py", line 825, in __get__
    record.ensure_one()
  File "/opt/odoo/common/openerp/v8/openerp/models.py", line 5355, in ensure_one
    raise except_orm("ValueError", "Expected singleton: %s" % self)
except_orm: ('ValueError', 'Expected singleton: ir.translation(4899, 703976)')
```

With this patch, now we let Odoo return the translated string by using its
standard method to do so, so we have to care for less.

@Tecnativa

yajo committed:
[FIX][base_export_manager] Fix "Expected singleton" bug.

If you had a field that got translated in more than 1 addon, you'd possibly get
to this error:

      File "/opt/odoo/0079_ahk_openerp/oca/base_export_manager/models/ir_exports_line.py", line 105, in _compute_label
        field.name)),
      File "/opt/odoo/common/openerp/v8/openerp/fields.py", line 825, in __get__
        record.ensure_one()
      File "/opt/odoo/common/openerp/v8/openerp/models.py", line 5355, in ensure_one
        raise except_orm("ValueError", "Expected singleton: %s" % self)
    except_orm: ('ValueError', 'Expected singleton: ir.translation(4899, 703976)')

With this patch, now we let Odoo return the translated string by using its
standard method to do so, so we have to care for less.

yajo commented:
Please could somebody forward-patch this to v9 when merged? Or do I open a separate PR?

yajo committed:
Move installation outside a data file.

This makes the whole installation to be able to roll back if something goes
wrong, instead of entering an error loop.

yajo commented:
Before https://github.com/OCA/server-tools/pull/521/commits/9ef4280b6c529cf36404365eb963cbe442deea29, if you had a buggy record in database, you got an endless error loop. Now you get this beautiful message:

![captura de pantalla de 2016-08-17 13-40-42](https://cloud.githubusercontent.com/assets/973709/17735128/ef0a5c32-6480-11e6-8569-dd4e9dad2209.png)

I considered adding a backup column of `ir.exports.line.name`, but it's not necessary because now installation gets rolled back if anything fails, and it does not lose data.

pedrobaeza commented on base_export_manager/hooks.py:
```diff
@@ -0,0 +1,19 @@
+    char field with stored values. We have to inverse it to avoid database
+    inconsistencies.
+    """
+    env = api.Environment(cr, SUPERUSER_ID, dict())
```
```
with api.Environment.manage():
    env = api.Environment(cr, SUPERUSER_ID, {})
```

pedrobaeza commented on base_export_manager/hooks.py:
```diff
@@ -0,0 +1,19 @@
+        ("field1_id", "=", False),
+        ("export_id", "!=", False),
+        ("name", "!=", False),
+    ])._inverse_name()
```
I still think we should protect this and don't stop. OK, now we are not in a loop, but forcing to fix one by one each of the possible old fields is an annoyance.

yajo replied:
But the option of leaving records in broken status is not very desirable too... :confused:

The only option then I can imagine is unlinking wrong records.

pedrobaeza replied:
They are broken, yes, but you can correct all of them in a whole in the corresponding menu, and the break is only when trying to use the corresponding export profile, so it's not a big problem nor breaks the DB.

pedrobaeza replied:
Also when migrating from previous versions, you get the loop in the update also (I'm just facing this right now).

yajo committed:
Include envorionment in its manager.

pedrobaeza commented:
One thing I'm facing: can it be that the textual representation of the fields (field `name`) have changed? It seems that previously you have `account_id / code` for representing a sub field, and now it's `account_id/code` (without spaces).

pedrobaeza commented on base_export_manager/models/ir_exports_line.py:
```diff
@@ -162,9 +143,17 @@ def _get_field_id(self, model, name):
+        result = self.env["ir.model.fields"].search(
             [("name", "=", name),
              ("model_id", "=", model.id)])
+        try:
```
Use `if not result.exists():` instead

pedrobaeza committed:
[IMP] base_export_manager: Several changes

pedrobaeza committed:
[IMP] base_export_manager: Add 4th field

yajo commented:
Just to clarify: https://github.com/OCA/server-tools/pull/521#issuecomment-240485437 was not a bug.

Is this now ready to review?

pedrobaeza commented:
Indeed it is

pedrobaeza commented:
:+1: the part I haven't done

pedrobaeza commented:
@max3903 can you check this, please?

yajo commented on base_export_manager/models/ir_exports_line.py:
```diff
@@ -53,104 +62,106 @@ def _default_model1_id(self):
+            self.label = ("%s (%s)" % ("/".join(parts), self.name)
+                          if parts and self.name else False)
+        except KeyError:
+            pass
```
I'd thank a little comment about why this is necessary.

dreispt replied:
I concur with @yajo: it's not obvious why the `pass` is needed, an explaining comment would be welcome.

yajo commented on base_export_manager/models/ir_exports_line.py:
```diff
@@ -53,104 +62,106 @@ def _default_model1_id(self):
+                continue
+            field_name = parts[num - 1]
+            model = self.model_n(num)
+            self.with_context(skip_check=True)[self.field_n(num, True)] = (
```
This skip trick also deserves a little comment.

yajo commented on base_export_manager/views/ir_exports_view.xml:
```diff
@@ -47,33 +47,25 @@
+                        <field name="model4_id" invisible="True"/>
                         <field name="label"/>
-                        <field name="name"/>
+                        <field name="name" readonly="0"/>
```
I'd prefer it if you remove the readonly from the model itself.

yajo commented on base_export_manager/tests/test_ir_exports_line.py:
```diff
@@ -34,3 +42,44 @@ def test_get_label_string(self):
+            'field2_id': self.field_parent_id.id,
+            'field3_id': self.field_name.id,
+        })
+        self.assertEqual(line.name, 'parent_id/parent_id/name')
```
Could you add the 4th field here? It's a simple patch and covers more cases.

yajo commented on base_export_manager/tests/test_ir_exports_line.py:
```diff
@@ -34,3 +42,44 @@ def test_get_label_string(self):
+    def test_inverse_name(self):
+        line = self.env['ir.exports.line'].create({
+            'export_id': self.export.id,
+            'name': 'parent_id/parent_id/name',
```
Could you add the 4th field here? It's a simple patch and covers more cases.

yajo commented:
I'm thinking that maybe a good improvement would be to give this addon its own menu item. Non-admins may want to manage this too, given what this addon ultimately does is make that task easier. Maybe it could be handled by a group setting. Not a requirement though.

lasley commented on base_export_manager/models/ir_exports_line.py:
```diff
@@ -53,104 +62,106 @@ def _default_model1_id(self):
+            self.label = ("%s (%s)" % ("/".join(parts), self.name)
+                          if parts and self.name else False)
+        except KeyError:
+            pass
```
IMO it would also be a little nicer if we guarded a little closer to the key error. We could then just sub the `pass` for a `return` to accomplish the same effect

yajo replied:
I like the idea.

lasley commented on base_export_manager/models/ir_exports_line.py:
```diff
@@ -53,104 +62,106 @@ def _default_model1_id(self):
+    @api.one
+    @api.depends("field3_id")
+    def _compute_model4_id(self):
+        """Get the related model for the third field."""
```
s/third/fourth

lasley commented on base_export_manager/models/ir_exports_line.py:
```diff
@@ -53,104 +62,106 @@ def _default_model1_id(self):
+                raise exceptions.ValidationError(
+                    _("Field '%s' already exists") % self.name)
+
+    @api.onchange('name')
```
Isn't an `api.multi` required for onchange?

yajo replied:
`@api.multi` is implicit if the method signature does not start by `self, cr, uid, ...`. Its advantage is that makes methods callable from old api too, but if that's not a requirement, you don't need to put it. I'm more in the side of always decorating, but certainly it's not required.

max3903 commented:
@pedrobaeza Sorry, I missed your message. What do you want me to check?

pedrobaeza commented:
Well, the idea is to see the improvements and give your bless.

max3903 commented:
:+1: LGTM

rafaelbn commented:
Functionaly tested :+1:

dreispt commented on base_export_manager/models/ir_exports_line.py:
```diff
@@ -53,104 +62,106 @@ def _default_model1_id(self):
 
-    @api.multi
-    @api.depends("field1_id", "field2_id", "field3_id")
+    @api.one
```
Isn't best to leave api.maulti here, since we have confirmation that Odoo intends to deprecated api.one ?

pedrobaeza replied:
I prefer `@api.one` while it exists. And I repeat that it's still not clear that it will be deprecated.

dreispt replied:
I don't see the point for the extra work when doing the port to v9.
And when asking to reconsider deprecation, all comments I had were against keeping it. So I don't see must space for any hopes on that.

pedrobaeza replied:
No, there's no work needed. `@api.one` is still in v9, v10, and counting.

dreispt replied:
Won't Lint report errors?

pedrobaeza replied:
No, they are only warnings. The PR in MQT was for removing the warning.

pedrobaeza commented:
@Yajo, can you attend last comments (including yours)?

yajo committed:
Move to api.multi, refactoring some stuff.

- Add some comments in complex parts.
- Rename `onchange_name` to `_onchange_name` (guidelines).
- Make `_compute_name`'s try block shorter and easier to understand.

yajo committed:
Allow R/W of name directly in model.

yajo committed:
Update tests to cover new behaviors.

pedrobaeza commented:
Merging as the problem in Travis fluctuates between builds.
